File: packages/widget/src/ui/home-screen.ts
// The Home screen — the first thing the panel shows.
//
// Mirrors the reference product's `ChatHome`: the merchant's greeting, one
// prominent "start a conversation" card, the most recent conversation with a
// way to see the rest, and the merchant's Common Questions.
//
// ── What this does NOT own ───────────────────────────────────────────────
//
// The hero header above it, and the Common Questions list itself. Both already
// exist (`ui/hero-header.ts`, `ui/common-questions.ts`), are already driven by
// published config, and are already tested. This screen arranges them; it does
// not re-implement either, and a second Common Questions renderer would be a
// second place for one console setting to be interpreted.

import { el, icon, ICONS } from './dom.js';
import { relativeTimeLabel } from './session-picker.js';
import { statusPill } from './session-status.js';
import type { ChatStatus } from '@dhaam-ccrm/core';
import type { ChatSessionSummary } from '@dhaam-ccrm/js';
import type { ResolvedEntry } from '../remote-config.js';

/**
 * Which statuses Home's "Recent conversation" section shows at all.
 *
 * `OPEN` and nothing else, decided by the product owner in answer to a direct
 * question: this section is a shortcut back into the conversation still
 * going, not a second conversation list. A row that is `RESOLVED` or `CLOSED`
 * is over; `WAITING_FOR_AGENT`, `ASSIGNED` and `ON_HOLD` are live but are
 * waiting on the merchant rather than on the customer, and putting either
 * kind at the top of Home invites a tap that leads somewhere the customer has
 * nothing left to do. A narrower rule than "hide the finished ones" was
 * offered alongside the broader one and this is the one that was chosen — so
 * widening it is a product decision, not a tidy-up.
 *
 * Nothing this table hides becomes unreachable: a conversation in any of the
 * four statuses it hides and still lists — `RESOLVED`, `WAITING_FOR_AGENT`,
 * `ASSIGNED`, `ON_HOLD` — is still on the Messages screen, and the bottom tab
 * bar that opens it is on screen whenever Home is. Not behind this section's
 * own "See all" — that button is a CHILD of the section (`recentSection`
 * below), so hiding the section takes "See all" with it. The Messages tab is
 * the whole surviving route, which is why it has to be one that is always
 * there.
 *
 * That reachability argument NARROWED on 2026-09-14, and only for `CLOSED`.
 * It used to read "every conversation in every status is still on the
 * Messages screen"; that is no longer true. `widget.ts` now withholds closed
 * conversations from the customer's surfaces upstream of this screen (see
 * `customerVisibleSessions` there), so a `CLOSED` conversation is on neither
 * Home nor Messages and has no surviving customer route at all — deliberately,
 * by user decision: closed is the merchant taking it off the table, and a
 * route back into something the customer can do nothing with is a dead end.
 * `RESOLVED`, `WAITING_FOR_AGENT`, `ASSIGNED` and `ON_HOLD` are unaffected and
 * the paragraph above holds for them in full, which is what matters when
 * weighing this table: hiding a status HERE still costs nothing but a tap for
 * every status except `CLOSED`, which is already withheld upstream — except
 * for the joined-session case, which is the one way a `CLOSED` summary still
 * reaches `update()` at all (see the `@param recent` note below).
 *
 * A `Record<ChatStatus, boolean>` rather than `status === 'OPEN'`, for exactly
 * the reason `ui/session-status.ts` gives for its own table: a seventh wire
 * status has to be a COMPILE ERROR here. `!== 'OPEN'` would silently keep a
 * new status hidden, which is the safe half; `=== 'OPEN'` written the other
 * way round would silently show it. Neither would make anyone decide.
 */
const SHOWN_IN_RECENT: Record<ChatStatus, boolean> = {
  OPEN: true,
  WAITING_FOR_AGENT: false,
  ASSIGNED: false,
  ON_HOLD: false,
  RESOLVED: false,
  CLOSED: false,
};

export interface HomeScreenCallbacks {
  /** Start a fresh conversation. */
  readonly onStartNew: () => void;
  /** Open the conversation named by id. */
  readonly onOpenConversation: (sessionId: string) => void;
  /** Go to the full conversation list. */
  readonly onSeeAll: () => void;
  /** Row 1's alt and row 2's CTA — hand the visitor to the web-form surface. */
  readonly onLeaveMessage: () => void;
  /** Row 2's alt — reach a real chat despite the tenant being closed. */
  readonly onChooseChat: () => void;
}

export interface HomeScreenView {
  readonly node: HTMLElement;
  /**
   * @param recent the newest of the conversations the customer still has
   *   (see below on the one status the caller withholds), or `null` when
   *   there is none — which is the case for every first-time visitor, and part of why
   *   the whole "Recent conversation" section is conditional rather than an
   *   empty-state box. The other part is its STATUS: the section shows only
   *   for an `OPEN` one (see {@link SHOWN_IN_RECENT}), so passing a resolved
   *   or waiting conversation here renders the row and hides the section.
   *
   *   Which conversation that is remains the caller's choice, not this
   *   screen's, and that choice CHANGED on 2026-09-14 by explicit user
   *   decision. It used to be the newest session overall, so a visitor whose
   *   newest conversation had just been closed saw no Recent section even
   *   when an older open one existed; that is superseded. `widget.ts` now
   *   filters `CLOSED` out of the customer's conversations
   *   (`customerVisibleSessions`) before this screen is handed anything, so
   *   `mostRecentSession` skips closed ones and a closed conversation no
   *   longer stands in front of an open one here.
   *
   *   With one exception, which is why this can still be handed a `CLOSED`
   *   summary: the conversation the customer is currently JOINED to is exempt
   *   from that filter (so it cannot vanish from under them mid-read — see
   *   `customerVisibleSessions`), and while it is both closed and newest it
   *   still wins `mostRecentSession` and hides this section even when an
   *   older OPEN conversation exists.
   *
   *   Only `CLOSED`. `mostRecentSession` is still not "the newest OPEN one":
   *   a visitor whose newest conversation has just been RESOLVED still sees
   *   no Recent section even when an older open one exists, because
   *   {@link SHOWN_IN_RECENT} — a separate, still-valid decision — shows the
   *   section for `OPEN` and nothing else. The older one is one tap away on
   *   Messages.
   * @param entry the resolved support entry — which of the six PRD rows this
   *   render is for. Drives the CTA's title/sub-line, the alt button beside
   *   it, and the mid-visit chat→ticket announcement.
   */
  update(recent: ChatSessionSummary | null, subtitle: string, entry: ResolvedEntry): void;
}

/**
 * The CTA's title and sub-line for `entry`.
 *
 * A three-entry table, keyed on `primary` — NOT a table of all six rows,
 * because the CTA only ever says one of three things. `source` is checked
 * FIRST and separately from `primary`: an `'assumed'` entry's `primary` is
 * always `'chat'` (see `entryFor`), and without this guard that would take
 * the 'chat' branch below and start showing "Chat now" — a claim of
 * knowledge — on every deployment that has not yet published a `support`
 * block, which is every deployment on the day this ships. Today's exact
 * copy, "Send us a message", is what an assumed entry gets instead.
 */
function ctaCopy(entry: ResolvedEntry, subtitle: string): { readonly title: string; readonly sub: string } {
  if (entry.source === 'published' && entry.primary === 'ticket') {
    return {
      title: 'Leave a message',
      // Row 2 (chat also on, but closed) vs row 5 (chat never offered): the
      // widget must not imply a live human is waiting either way, but row 2
      // alone gets to say the team is closed, because only there was a live
      // option ever on the table.
      sub: entry.secondary === 'chat' ? "We're closed — we'll reply by email." : "We'll reply by email.",
    };
  }
  if (entry.source === 'published' && entry.primary === 'chat') {
    return { title: 'Chat now', sub: subtitle };
  }
  return { title: 'Send us a message', sub: subtitle };
}

/** The alt button beneath the CTA, or `null` for a row that offers none. */
function altFor(entry: ResolvedEntry): { readonly label: string; readonly kind: 'ticket' | 'chat' } | null {
  if (entry.source !== 'published') return null;
  if (entry.secondary === 'ticket') return { label: 'Leave a message instead', kind: 'ticket' };
  if (entry.secondary === 'chat') return { label: 'Try live chat anyway', kind: 'chat' };
  return null;
}

export function createHomeScreen(callbacks: HomeScreenCallbacks): HomeScreenView {
  // The primary action. A card rather than a button, because it carries a
  // second line — the response-time promise the merchant configured — and a
  // button with two lines of different weight is a card that has not admitted
  // it yet.
  const ctaSubtitle = el('span', { attrs: { class: 'dh-home-cta-sub' } });
  const ctaTitle = el('span', { attrs: { class: 'dh-home-cta-title' }, text: 'Send us a message' });
  // Reassigned on every `update()`, same as `recentRow.onclick` below —
  // rebuilding the button itself would take it out from under a visitor
  // mid-press, the same reason the launcher glyph lives in a slot.
  let onCtaPress: () => void = callbacks.onStartNew;
  const cta = el('button', {
    attrs: { class: 'dh-home-cta', type: 'button' },
    children: [
      el('span', { attrs: { class: 'dh-home-cta-icon' }, children: [icon(ICONS.chat, 20)] }),
      el('span', { attrs: { class: 'dh-home-cta-text' }, children: [ctaTitle, ctaSubtitle] }),
      el('span', { attrs: { class: 'dh-home-chevron', 'aria-hidden': 'true' }, text: '›' }),
    ],
    on: { click: () => onCtaPress() },
  });

  // The alt affordance beside the CTA — Row 1's "Leave a message instead" or
  // Row 2's "Try live chat anyway". A real `<button>` with a visible text
  // label, not an `<a>` (it navigates nowhere).
  //
  // Its own class, deliberately NOT `.dh-form-skip`: that class already means
  // something specific in every test and every other surface — "the one Skip
  // /Cancel control in the currently-open form" — and this button sits on
  // Home, permanently in the DOM (hidden, not removed) whichever screen is
  // showing. Sharing the class made `.dh-form-skip` stop being unique the
  // moment a surface with its own Skip button was open at the same time,
  // which broke an existing, unrelated pre-chat test that queries for
  // exactly that uniqueness. `.dh-home-alt` gets `.dh-form-skip`'s LOOK
  // (styles.ts) without its identity.
  let onAltPress: () => void = () => undefined;
  const alt = el('button', {
    attrs: { class: 'dh-home-alt', type: 'button', hidden: true },
    on: { click: () => onAltPress() },
  });

  // The mid-visit chat→ticket announcement — `role="status"`, never `alert`,
  // for the reason `ui/offline-banner.ts` already gives about interrupting a
  // screen reader mid-sentence. Hidden until a flip is actually observed.
  const notice = el('p', {
    attrs: { class: 'dh-entry-note', role: 'status', 'aria-live': 'polite', hidden: true },
  });

  // ── Recent conversation ────────────────────────────────────────────────
  const recentTitle = el('span', { attrs: { class: 'dh-home-recent-title' } });
  const recentStatus = el('span', { attrs: { class: 'dh-home-recent-status' } });
  const recentPreview = el('span', { attrs: { class: 'dh-home-recent-preview' } });
  const recentTime = el('span', { attrs: { class: 'dh-home-recent-time' } });
  const recentRow = el('button', {
    attrs: { class: 'dh-home-recent-row', type: 'button' },
    children: [
      el('span', {
        attrs: { class: 'dh-home-recent-body' },
        children: [
          el('span', { attrs: { class: 'dh-home-recent-head' }, children: [recentTitle, recentStatus] }),
          recentPreview,
          recentTime,
        ],
      }),
      el('span', { attrs: { class: 'dh-home-chevron', 'aria-hidden': 'true' }, text: '›' }),
    ],
  });

  const seeAll = el('button', {
    attrs: { class: 'dh-home-seeall', type: 'button' },
    text: 'See all',
    on: { click: () => callbacks.onSeeAll() },
  });

  const recentSection = el('section', {
    attrs: { class: 'dh-home-section', hidden: true, 'aria-labelledby': 'dh-home-recent-heading' },
    children: [
      el('div', {
        attrs: { class: 'dh-home-section-head' },
        children: [
          el('h3', {
            attrs: { class: 'dh-home-section-title', id: 'dh-home-recent-heading' },
            text: 'Recent conversation',
          }),
          seeAll,
        ],
      }),
      recentRow,
    ],
  });

  // The Common Questions row is MOUNTED here by the widget, not built here —
  // it is the same `commonQuestionsHost` the conversation screen used, moved
  // rather than duplicated.
  const questionsSlot = el('section', {
    attrs: { class: 'dh-home-section dh-home-questions', hidden: true, 'aria-labelledby': 'dh-home-q-heading' },
    children: [
      el('h3', {
        attrs: { class: 'dh-home-section-title', id: 'dh-home-q-heading' },
        text: 'Common Questions',
      }),
    ],
  });

  const node = el('div', {
    attrs: { class: 'dh-home' },
    children: [notice, cta, alt, recentSection, questionsSlot],
  });

  // `null` until the first `update()` — see the flip guard below, which must
  // not fire on the very first render just because it differs from an
  // arbitrary seed.
  let previousPrimary: ResolvedEntry['primary'] | null = null;

  return {
    node,
    update(recent, subtitle, entry) {
      // The merchant's own response-time line, reused rather than a second
      // hardcoded "We usually reply instantly" — it is the same promise the
      // status line makes, and two copies would drift. Used only when the
      // CTA is actually about chat; `ctaCopy` supplies its own sentence for
      // a ticket row.
      const copy = ctaCopy(entry, subtitle);
      ctaTitle.textContent = copy.title;
      ctaSubtitle.textContent = copy.sub;
      ctaSubtitle.hidden = copy.sub === '';
      onCtaPress =
        entry.source === 'published' && entry.primary === 'ticket'
          ? callbacks.onLeaveMessage
          : callbacks.onStartNew;

      const altChoice = altFor(entry);
      alt.hidden = altChoice === null;
      if (altChoice !== null) {
        alt.textContent = altChoice.label;
        onAltPress = altChoice.kind === 'ticket' ? callbacks.onLeaveMessage : callbacks.onChooseChat;
      }

      // Only a genuine chat→ticket flip announces itself, and only once per
      // flip — a repaint that lands on the same row again (a second message
      // arriving, a session change) must not restart the announcement mid-
      // sentence, the same discipline `ui/offline-banner.ts` applies to its
      // own live region.
      const flipped =
        previousPrimary === 'chat' && entry.source === 'published' && entry.primary === 'ticket';
      if (flipped) {
        const text = "We've just closed — leave a message and we'll reply by email.";
        if (notice.textContent !== text) notice.textContent = text;
        notice.hidden = false;
      } else {
        notice.hidden = true;
      }
      previousPrimary = entry.source === 'published' ? entry.primary : 'chat';

      // Hidden unless there IS one and it is still OPEN — see
      // {@link SHOWN_IN_RECENT}. The row's own content is still filled in
      // below for every status, so this is a visibility decision and not an
      // emptying one: the pill's words stay correct, and a future rule that
      // shows more statuses needs no second change here.
      recentSection.hidden = recent === null || !SHOWN_IN_RECENT[recent.status];
      if (recent === null) return;

      // NOT a subject line. The reference product shows one ("Delivery
      // issue", "Refund request"), but that is its own mock data: there is no
      // subject, title or topic on `chat_sessions`, in the REST projection, or
      // on core's `ChatSessionSummary`. Inventing one — from the first message,
      // say — would put a label on the row that the customer never wrote and
      // that nothing else in the product agrees with.
      //
      const s = recent as any;
      recentTitle.textContent =
        s.storeName?.trim() ||
        s.merchantName?.trim() ||
        recent.handledBy?.displayName ||
        s.adminName?.trim() ||
        'Support';
      // ALWAYS a pill, for every status. This used to carry a private
      // three-status table (RESOLVED/CLOSED/WAITING_FOR_AGENT) and render
      // nothing at all for OPEN, ASSIGNED and ON_HOLD — so the conversation a
      // customer was most likely still in the middle of was the one row that
      // refused to say where it stood, which is the half of defect 4 that is
      // purely client-side. The words come from `ui/session-status.ts`, the
      // same table the Messages list reads, so the two screens cannot drift.
      recentStatus.textContent = statusPill(recent.status);
      recentStatus.setAttribute('data-status', recent.status);
      const previewText = recent.lastMessagePreview || s.subject || '';
      recentPreview.textContent = previewText;
      recentPreview.hidden = previewText === '';
      recentTime.textContent = relativeTimeLabel(recent.lastMessageAt ?? recent.createdAt);
      recentRow.onclick = () => callbacks.onOpenConversation(recent.id);
    },
  };
}

/** Where the widget mounts the shared Common Questions row on this screen. */
export function homeQuestionsSlot(view: HomeScreenView): HTMLElement {
  return view.node.querySelector('.dh-home-questions')!;
}

File: packages/flutter/lib/src/ui/home_screen.dart
/// The Home screen — the first thing a customer sees. Mirrors
/// `ui/home-screen.ts`: the hero, a "Send us a message" CTA card, the most
/// recent conversation (when there is one) with a See-all link, and Common
/// Questions.
///
/// Reads [ChatWidgetCubit]'s state directly via [BlocBuilder] rather than
/// taking typed props — this is the SCREEN, the thing the root widget
/// mounts, not a presentational component a test builds in isolation (those
/// are [HeroHeader] and [CommonQuestionsList], which this screen composes).
library;

import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

import '../config/remote_config.dart';
import '../nav/chat_screens.dart';
import '../session/chat_session_summary.dart';
import '../session/session_display.dart';
import '../state/chat_widget_cubit.dart';
import '../state/chat_widget_state.dart';
import '../theme/chat_theme.dart';
import 'common_questions_list.dart';
import 'hero_header.dart';

/// The accessible name of the close control, on both screens that carry one.
///
/// ── One name, because it is one control ─────────────────────────────────
///
/// Home and Messages are tabs, not two places: the customer sees the same
/// affordance in the same corner whichever one they are on, and a screen
/// reader that announced two different names would describe it as two
/// different controls. So `messages_screen.dart` imports exactly this one
/// name (`show kCloseChatLabel`) rather than restating the string — one
/// constant here, no widget shared, no layout borrowed.
///
/// It is the accessible NAME and not decoration: `IconButton` forwards
/// `tooltip` to the semantics node, which is how the ⋯ menu
/// ('Conversation options') and the session switcher already get theirs.
/// Left unset, an icon-only button has no name at all.
const String kCloseChatLabel = 'Close chat';

/// The close control `ChatWidget.onClose` backs.
///
/// Right-aligned above the screen's own content, so it lands where a panel's
/// dismiss control is looked for and never on top of something it would
/// obscure.
///
/// The callback is REQUIRED and non-null here: "there is nobody to tell" is
/// answered by not building this at all, at the one place that knows it —
/// the screen's own `build`. A widget that rendered a zero-sized box for a
/// null callback would still be a widget in a tree that is supposed to be
/// unchanged, and a DISABLED button would be worse: a control that looks
/// like a way out and is not.
class ChatCloseButton extends StatelessWidget {
  const ChatCloseButton({super.key, required this.onClose});

  /// Invoked once per press, with no navigation of this package's own — see
  /// `ChatWidget.onClose` on why the pop is the host's.
  final VoidCallback onClose;

  @override
  Widget build(BuildContext context) {
    return Align(
      alignment: Alignment.centerRight,
      child: Padding(
        padding: const EdgeInsets.fromLTRB(4, 4, 4, 0),
        child: IconButton(
          onPressed: onClose,
          tooltip: kCloseChatLabel,
          icon: const Icon(Icons.close),
        ),
      ),
    );
  }
}

class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key, this.onClose});

  /// Forwarded from `ChatWidget.onClose`. Null draws no close control, and
  /// this screen is then the very tree it was before the parameter existed —
  /// see the `build` below, which does not even wrap the hero.
  final VoidCallback? onClose;

  @override
  Widget build(BuildContext context) {
    return BlocBuilder<ChatWidgetCubit, ChatWidgetState>(
      builder: (BuildContext context, ChatWidgetState state) {
        final ChatWidgetCubit cubit = context.read<ChatWidgetCubit>();
        final RemoteConfig config = state.config;
        // The most recent conversation the customer STILL HAS — closed ones
        // are not among them (see
        // [ChatWidgetState.customerVisibleSessions]).
        //
        // That is a decision the user made directly, and it is why the
        // filter is applied BEFORE `mostRecentSummary` rather than after:
        // asked whether a customer whose newest conversation had just been
        // closed should see no Recent section at all or the newest one they
        // can still do something with, they chose the second. Filtering
        // afterwards — `mostRecentSummary` first, then dropping a closed
        // winner — would produce the first, with an older open conversation
        // sitting unmentioned behind a section that had vanished.
        //
        // Only CLOSED falls through this way. This screen has no
        // status-gating table (unlike `home-screen.ts`'s `SHOWN_IN_RECENT`):
        // every status it is handed renders its row and its pill, so a
        // RESOLVED or WAITING_FOR_AGENT newest still leads Home exactly as
        // it did.
        final ChatSessionSummary? recent =
            mostRecentSummary(state.customerVisibleSessions);
        final double radius = chatCornerRadius(config);

        // ── The hero is pinned ABOVE the scroll view, not inside it ──
        //
        // Collapsing has to hand the hero's height back to something. A hero
        // that were merely the scroll view's first child would already be
        // gone by the time the visitor had scrolled past it, so there would
        // be nothing for a collapse to mean; pinned above, it stays put while
        // the content moves under it and then gets out of the way entirely.
        // `CollapsingHeroHeader` owns that arrangement rather than this
        // screen assembling it — see its own header for why the band and the
        // scroll view have to be one widget's business.
        final Widget hero = CollapsingHeroHeader(
          config: config,
          child: ListView(
            padding: EdgeInsets.zero,
            children: <Widget>[
              Padding(
                padding: const EdgeInsets.all(20),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: <Widget>[
                    _SendMessageCta(
                      // The merchant's own response-time line, reused rather
                      // than a second hardcoded "we usually reply instantly" —
                      // home-screen.ts's own comment on why this is the SAME
                      // ctaSubtitle the hero's own CTA would have shown, now
                      // that this card is the one place it renders (see
                      // hero_header.dart's header).
                      subtitle: config.header.ctaSubtitle,
                      radius: radius,
                      onTap: cubit.startNewConversation,
                    ),
                    if (recent != null) ...<Widget>[
                      const SizedBox(height: 24),
                      _RecentConversationSection(
                        summary: recent,
                        radius: radius,
                        onSeeAll: () => cubit.switchTab(ScreenName.messages),
                        onOpen: () => cubit.openConversation(recent.id),
                      ),
                    ],
                    if (config.commonQuestions.isNotEmpty) ...<Widget>[
                      const SizedBox(height: 24),
                      _SectionHeading('Common Questions'),
                      const SizedBox(height: 8),
                      CommonQuestionsList(
                        questions: config.commonQuestions,
                        // One call, not "open the new-conversation form and
                        // then send into it": a tapped question is a customer
                        // asking one specific thing, and routing it through the
                        // form would put the merchant's pre-chat questions in
                        // front of an answer they already asked for. See
                        // ChatWidgetCubit.startCommonQuestion.
                        onSelect: cubit.startCommonQuestion,
                      ),
                    ],
                  ],
                ),
              ),
            ],
          ),
        );

        // ── The close control, and the tree a host without one gets ──
        //
        // Returned UNWRAPPED when there is no callback: no Column, no
        // Expanded, nothing between the hero and this screen's caller. A
        // host that passes no `onClose` therefore gets the identical tree it
        // got before the parameter existed, which is the claim
        // `close_affordance_test.dart` makes and the reason this is a branch
        // rather than an `if` inside a Column that would always be built.
        //
        // Above the hero rather than over it. The hero is a merchant-coloured
        // band that renders nothing at all for a tenant who configured none
        // (see `HeroHeader`), so an icon floated on top of it would have no
        // guaranteed background to be legible against — and it COLLAPSES on
        // scroll, which would take the way out with it.
        final VoidCallback? close = onClose;
        if (close == null) return hero;
        return Column(
          children: <Widget>[
            ChatCloseButton(onClose: close),
            Expanded(child: hero),
          ],
        );
      },
    );
  }
}

class _SectionHeading extends StatelessWidget {
  const _SectionHeading(this.text);

  final String text;

  @override
  Widget build(BuildContext context) {
    return Text(
      text.toUpperCase(),
      style: Theme.of(context).textTheme.labelMedium?.copyWith(
            color: Theme.of(context).colorScheme.onSurfaceVariant,
            letterSpacing: 0.4,
          ),
    );
  }
}

class _SendMessageCta extends StatelessWidget {
  const _SendMessageCta(
      {required this.subtitle, required this.radius, required this.onTap});

  /// `config.header.ctaSubtitle` — hidden when the merchant left it unset,
  /// matching `home-screen.ts`'s own `ctaSubtitle.hidden = subtitle === ''`.
  final String? subtitle;
  final double radius;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final ColorScheme scheme = Theme.of(context).colorScheme;
    return Material(
      color: scheme.surface,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(radius),
        side: BorderSide(color: scheme.outlineVariant),
      ),
      child: InkWell(
        borderRadius: BorderRadius.circular(radius),
        onTap: onTap,
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Row(
            children: <Widget>[
              CircleAvatar(
                backgroundColor: scheme.primary,
                foregroundColor: scheme.onPrimary,
                child: const Icon(Icons.chat_bubble_outline, size: 20),
              ),
              const SizedBox(width: 12),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Text('Send us a message',
                        style: Theme.of(context).textTheme.titleSmall),
                    if (subtitle != null && subtitle!.isNotEmpty)
                      Text(
                        subtitle!,
                        style: Theme.of(context)
                            .textTheme
                            .bodySmall
                            ?.copyWith(color: scheme.onSurfaceVariant),
                      ),
                  ],
                ),
              ),
              Icon(Icons.chevron_right, color: scheme.onSurfaceVariant),
            ],
          ),
        ),
      ),
    );
  }
}

class _RecentConversationSection extends StatelessWidget {
  const _RecentConversationSection({
    required this.summary,
    required this.radius,
    required this.onSeeAll,
    required this.onOpen,
  });

  final ChatSessionSummary summary;
  final double radius;
  final VoidCallback onSeeAll;
  final VoidCallback onOpen;

  @override
  Widget build(BuildContext context) {
    final ColorScheme scheme = Theme.of(context).colorScheme;
    // ALWAYS a pill, for every status — see `homeStatusPill`. This used to be
    // a nullable map lookup that rendered nothing for OPEN, ASSIGNED and
    // ON_HOLD.
    final String pill = homeStatusPill(summary.status);
    final String preview = summary.lastMessagePreview ?? '';

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: <Widget>[
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            _SectionHeading('Recent conversation'),
            TextButton(onPressed: onSeeAll, child: const Text('See all')),
          ],
        ),
        Material(
          color: scheme.surface,
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(radius),
            side: BorderSide(color: scheme.outlineVariant),
          ),
          child: InkWell(
            borderRadius: BorderRadius.circular(radius),
            onTap: onOpen,
            child: Padding(
              padding: const EdgeInsets.all(16),
              child: Row(
                children: <Widget>[
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: <Widget>[
                        Row(
                          children: <Widget>[
                            Flexible(
                              child: Text(
                                // NOT a subject line — see chat_session_summary.dart's
                                // header on why this is WHO handled it, never an
                                // invented title. handledByText is not reused here
                                // because it prefixes "with " for the row-body
                                // context Messages uses; this is the row's own
                                // heading, so the bare name is what home-screen.ts
                                // itself renders here.
                                summary.handledBy?.displayName ??
                                    'Conversation',
                                overflow: TextOverflow.ellipsis,
                                style: Theme.of(context).textTheme.titleSmall,
                              ),
                            ),
                            const SizedBox(width: 8),
                            _StatusPill(pill),
                          ],
                        ),
                        if (preview.isNotEmpty)
                          Padding(
                            padding: const EdgeInsets.only(top: 2),
                            child: Text(
                              preview,
                              maxLines: 1,
                              overflow: TextOverflow.ellipsis,
                              style: Theme.of(context)
                                  .textTheme
                                  .bodySmall
                                  ?.copyWith(color: scheme.onSurfaceVariant),
                            ),
                          ),
                        Padding(
                          padding: const EdgeInsets.only(top: 2),
                          child: Text(
                            relativeTimeLabel(
                                summary.lastMessageAt ?? summary.createdAt),
                            style: Theme.of(context)
                                .textTheme
                                .bodySmall
                                ?.copyWith(color: scheme.onSurfaceVariant),
                          ),
                        ),
                      ],
                    ),
                  ),
                  Icon(Icons.chevron_right, color: scheme.onSurfaceVariant),
                ],
              ),
            ),
          ),
        ),
      ],
    );
  }
}

class _StatusPill extends StatelessWidget {
  const _StatusPill(this.label);

  final String label;

  @override
  Widget build(BuildContext context) {
    final ColorScheme scheme = Theme.of(context).colorScheme;
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
      decoration: BoxDecoration(
          color: scheme.surfaceContainerHighest,
          borderRadius: BorderRadius.circular(999)),
      child: Text(label,
          style: Theme.of(context)
              .textTheme
              .labelSmall
              ?.copyWith(color: scheme.onSurfaceVariant)),
    );
  }
}
